File: crates/medousa-sdk/src/jobs.rs
#[cfg(feature = "async")]
use medousa_types::{
    ArchiveAskJobRequest, ArchiveAskJobResponse, AskJobCompleteActionsRequest,
    AskJobCompleteActionsResponse, EnqueueAskRequest, EnqueuePromptRequest, EnqueueReportRequest,
    EnqueueResponse, JobReportResponse, JobResultResponse,
};

#[cfg(feature = "async")]
use crate::client::MedousaClient;
use crate::generated::ops;
use crate::op::op_path;
#[cfg(feature = "async")]
use crate::transport::decode;

#[cfg(feature = "async")]
pub struct JobsApi<'a> {
    pub(crate) client: &'a MedousaClient,
}

#[cfg(feature = "async")]
impl JobsApi<'_> {
    pub async fn enqueue_ask(
        &self,
        request: &EnqueueAskRequest,
    ) -> Result<EnqueueResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), ops::JOBS_ASK_POST.path, body)
            .await?;
        decode(value).await
    }

    pub async fn result(&self, job_id: &str) -> Result<JobResultResponse, crate::SdkError> {
        let path = op_path(
            &ops::JOBS_BY_JOB_ID_RESULT_GET,
            &[("job_id", job_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn report(&self, job_id: &str) -> Result<JobReportResponse, crate::SdkError> {
        let path = op_path(
            &ops::JOBS_BY_JOB_ID_REPORT_GET,
            &[("job_id", job_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), &path)
            .await?;
        decode(value).await
    }

    pub async fn enqueue_report(
        &self,
        request: &EnqueueReportRequest,
    ) -> Result<EnqueueResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), ops::JOBS_REPORT_POST.path, body)
            .await?;
        decode(value).await
    }

    pub async fn enqueue_prompt(
        &self,
        request: &EnqueuePromptRequest,
    ) -> Result<EnqueueResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), ops::JOBS_PROMPT_POST.path, body)
            .await?;
        decode(value).await
    }

    pub async fn complete_actions(
        &self,
        job_id: &str,
        request: &AskJobCompleteActionsRequest,
    ) -> Result<AskJobCompleteActionsResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::JOBS_BY_JOB_ID_COMPLETE_ACTIONS_POST,
            &[("job_id", job_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }

    pub async fn archive(
        &self,
        job_id: &str,
        request: &ArchiveAskJobRequest,
    ) -> Result<ArchiveAskJobResponse, crate::SdkError> {
        let body =
            serde_json::to_value(request).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let path = op_path(
            &ops::JOBS_BY_JOB_ID_ARCHIVE_POST,
            &[("job_id", job_id.trim())],
        )?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), &path, body)
            .await?;
        decode(value).await
    }
}

File: python/medousa-sdk/src/medousa/jobs.py
from __future__ import annotations

from medousa._decode import decode
from medousa._ops import op_path
from medousa.client import MedousaClient
from medousa.types import (
    ArchiveAskJobRequest,
    ArchiveAskJobResponse,
    AskJobCompleteActionsRequest,
    AskJobCompleteActionsResponse,
    EnqueueAskRequest,
    EnqueuePromptRequest,
    EnqueueReportRequest,
    EnqueueResponse,
    JobReportResponse,
    JobResultResponse,
)


class JobsApi:
    def __init__(self, client: MedousaClient) -> None:
        self._client = client

    async def enqueue_ask(self, request: EnqueueAskRequest) -> EnqueueResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("jobs.ask.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(EnqueueResponse, value)

    async def result(self, job_id: str) -> JobResultResponse:
        value = await self._client.transport.get_json(
            self._client.base_url,
            op_path("jobs.by_job_id.result.get", job_id=job_id),
        )
        return decode(JobResultResponse, value)

    async def report(self, job_id: str) -> JobReportResponse:
        value = await self._client.transport.get_json(
            self._client.base_url,
            op_path("jobs.by_job_id.report.get", job_id=job_id),
        )
        return decode(JobReportResponse, value)

    async def enqueue_report(self, request: EnqueueReportRequest) -> EnqueueResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("jobs.report.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(EnqueueResponse, value)

    async def enqueue_prompt(self, request: EnqueuePromptRequest) -> EnqueueResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("jobs.prompt.post"),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(EnqueueResponse, value)

    async def complete_actions(
        self,
        job_id: str,
        request: AskJobCompleteActionsRequest,
    ) -> AskJobCompleteActionsResponse:
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("jobs.by_job_id.complete_actions.post", job_id=job_id),
            request.model_dump(mode="json", exclude_none=True),
        )
        return decode(AskJobCompleteActionsResponse, value)

    async def archive(
        self,
        job_id: str,
        request: ArchiveAskJobRequest | None = None,
    ) -> ArchiveAskJobResponse:
        body = (request or ArchiveAskJobRequest()).model_dump(mode="json", exclude_none=True)
        value = await self._client.transport.post_json(
            self._client.base_url,
            op_path("jobs.by_job_id.archive.post", job_id=job_id),
            body,
        )
        return decode(ArchiveAskJobResponse, value)
